### PR TITLE
hist.Hist.fill method performance boost

### PR DESCRIPTION
**Pull request by pfackeldey**
Dear coffea developers,

this PR basically substitutes the `np.add.at` DenseAxis filling with `np.bincount`. This results in a significant performance boost:
image (1)
Here the `fill` method is the current version, while the `fast_fill` method is the `np.bincount` implementation (this method naming is just for the performance showcase).

Real world performance gain:
In our full analysis setup we are processing chunks of 100k events and fill O(100k) 1D-histograms per chunk. With the new `fill` (with `np.bincount`) we were able to speed up the filling of those histograms from ~80seconds to ~5seconds (while selection, reconstruction, etc. is just about ~10 seconds) per chunk. This significantly improves the runtime for a analysis at scale!
Have a look at our performance measures in our analysis case ("Histogramming").
with  `np.add.at`:
image (2)
with `np.bincount`:
image (3)

This implementation is also very close to the `np.histogramdd` implementation: https://github.com/numpy/numpy/blob/v1.19.0/numpy/lib/histograms.py#L942-L1128

I know this PR affects a very central method, which under no circumstances should break. I've made plenty of local tests successfully, also the unittests were successful. Therefore I'd like to request a careful review, especially regarding corner cases, which I might not have catched/tested.

One downside: The values, which are filled into the DenseAxes have to be of type `np.ndarray`! I've put an Exception to catch this in the code.

Thank you very much for reviewing!
Best, Peter

### CODE REVIEW

**Comment by lgray** [comment]
Excellent work! Thanks for the contribution. We'll figure out a proper review soon.

**Comment by nsmith-** [comment]
Interesting, I seem to recall someone (Henry?) comparing the options and finding add.at to be the best, but clearly not any more, or the tests were not applicable. This reminds me that, really, we should migrate to boost-histogram. I believe the last of the missing functionality was added as of v0.8, so we should check it out.

**Comment by lgray** [comment]
@nsmith- I see a decent definition of a 0.7. Awkward1 + boost-histogram transition.

File: coffea/hist/hist_tools.py [hunk 1]
```diff
@@ -963,14 +963,28 @@ def fill(self, **values):
                 self._sumw2[sparse_key] = np.zeros(shape=self._dense_shape, dtype=self._dtype)
 
         if self.dense_dim() > 0:
-            dense_indices = tuple(d.index(values[d.name]) for d in self._axes if isinstance(d, DenseAxis))
+            dense_indices = ()
+            for d in self._axes:
+                if isinstance(d, DenseAxis):
+                    if not isinstance(values[d.name], np.ndarray):
```
Comment (nsmith-) [inline coffea/hist/hist_tools.py]: You could consider instead to try `numpy.asarray(values[d.name])` to give a chance for compatible arrays to be cast to numpy. What is not clear to me is where integer arguments are rejected. What numpy function didn't accept them?

**Review by nsmith- — CHANGES_REQUESTED** [review]

File: coffea/hist/hist_tools.py (hunk 1 above)
Comment (lgray) [reply coffea/hist/hist_tools.py]: I was about to make the same comment. We should preserve original behavior, and it doesn't look like it is that hard.

Comment (lgray) [reply coffea/hist/hist_tools.py]: It's np.bincount - it expects array of dimension 1 exactly (and single numbers are dim 0).

Comment (nsmith-) [reply coffea/hist/hist_tools.py]: Ah, so the solution is to wrap the `xy` indexer with `np.atleast_1d`

Comment (pfackeldey) [reply coffea/hist/hist_tools.py]: Sure, good idea :) I can turn `values[d.name]` into a 1d array in case it is a list or a single digit with `np.atleast_1d`.

**Review by nsmith- — APPROVED** [review]
Thanks, this looks good

File: coffea/hist/hist_tools.py [hunk 2]
```diff
@@ -964,13 +964,22 @@ def fill(self, **values):
 
         if self.dense_dim() > 0:
             dense_indices = tuple(d.index(values[d.name]) for d in self._axes if isinstance(d, DenseAxis))
+            xy = np.atleast_1d(np.ravel_multi_index(dense_indices, self._dense_shape))
```
Comment (nsmith-) [inline coffea/hist/hist_tools.py:967]: A micro-optimisation would be to re-use the dense_indices variable so the input arrays can be deallocated
